src/FreqWindow.cpp

```diff
@@ -1013,9 +1018,32 @@ void FrequencyPlotDialog::SendRecalcEvent()
    GetEventHandler()->AddPendingEvent(e);
 }
 
+namespace {
+// Doesn't zero the buffer past `GetNumSamples(tracks)`.
+void FillBuffer(
+   float* buffer, size_t bufferSize,
+   const std::vector<ChannelGroupSampleView>& tracks)
+{
+   size_t numWritten = 0;
+   for (const auto& track : tracks)
+      for (const auto& channel : track)
+         for (const auto& segment : channel)
+         {
+            const auto toRead = limitSampleBufferSize(
+               bufferSize - numWritten, segment.GetSampleCount());
+            segment.Copy(buffer + numWritten, toRead);
+            numWritten += toRead;
+            if (numWritten == bufferSize)
+               return;
```

Comment:
This bothers me, for the same reasons that Mixer Board changes did, and why I cautioned about correct RMS calculation in Contrast.cpp.

Samples from neighboring clips with different stretches are treated as just one long sequence of samples as if they had the same real-time rate, ignoring stretches.  Is that right?

For Contrast I argued it was wrong:  rms should be calculated for clips separately, then combined into an overall RMS with the proper weighting of different clips, taking stretches into account at that point.

For Mixer Board, it was all right if only peak values are needed, but as I have also mentioned, Mixer Board also does an rms calculation that can be wrong now, but fixing it is future work that needs a totally different approach anyway to get the true post-fader levels out of the audio engine.

What wrong calculation could be happening here? …

Reply:
… answering that would take you into the guts of SpectrumAnalyst, where as I recall, values derived from many different FFT windows are combined into a single window that is later plotted.

As we have lately discussed, it is not wrong to take FFT of pre-stretched data and assign it to the same frequency bins as for the as-rendered sound that we don’t bother to calculate just yet.  However…

I think you would find there an assumption that all windows should be weighted equally in the combined result, but that could now be wrong with stretches.

What is a test case that might demonstrate this? …

Reply:
### TL, DR:
From a result point of view, IMO it very marginally matters whether we pre-render or not, because of how rare use cases are where the limitation of either approach becomes a problem. (I assume that we don't put in the effort of weighting according to stretch ratio.) Implementation wise, I think this one is easier, isn't it? No need to make an editable copy of the track objects, just use a different, user-friendly API (`GetSampleView`).

### Detail:
Yes, the weighting is ignored here, which might yield unexpected results if the selection is made across clips of different stretch ratios. I don't expect this to happen often, and if it does, that the stretching difference is large enough to be noticeable, but in principle, yes, either accurate, preliminary stretch-rendering, or weighting would alleviate this.

I said "accurate" because, as we know already, the stretching algorithm is not perfect. I've done an experiment, which should serve as material for the spectrum analysis as well as the noise reduction problem.

1. I recorded myself saying "test" with an open mic.
2. I stretched the clip moderately, squeezing it by approx. 25%.
3. I duplicated the track for better comparison.
4. I selected a short region of background noise,
5. I rendered the stretching of that region.

The project overall:
![BackgroundNoise project](https://github.com/audacity/audacity/assets/22740106/c2677a6d-f528-4f21-b315-0f4e51e34f73)

The spectrum analysis of the unrendered labeled segment (upper track):
![Background noise spectrum](https://github.com/audacity/audacity/assets/22740106/66d7a635-ce0b-4936-980b-07e005b0cf5a)

That of the rendered segment (lower track):
![Background noise spectrum - rendered](https://github.com/audacity/audacity/assets/22740106/f838b753-454b-491f-b0c3-65e231b2cb3a)

(A way of superposing the spectra would be nice, like in Adobe Audition. We don't have this yet, do we?)

For the spectrum analysis use case, I don't expect this to happen often either. It's more for the noise reduction case that this might become an issue, but I'll elaborate in that other PR.